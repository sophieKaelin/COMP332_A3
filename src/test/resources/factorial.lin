//
// This file is part of COMP332 Assignment 3 2019.
//
// Lintilla, a simple functional programming language.
//
// Â© 2019, Dominic Verity and Anthony Sloane, Macquarie University.
//         All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// A simple factorial function written in Lintilla.
//

fn fact(n: int) -> int {
  if n = 0 { 1 } else { n * fact(n-1) }
};

print(fact(5))