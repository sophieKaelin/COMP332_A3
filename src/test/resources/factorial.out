Program tree:
-------------

Program(
    Vector(
        FnDecl(
            IdnDef("fact"),
            Vector(ParamDecl(IdnDef("n"), IntType())),
            Some(IntType()),
            Block(
                Vector(
                    IfExp(
                        EqualExp(IdnExp(IdnUse("n")), IntExp(0)),
                        Block(Vector(IntExp(1))),
                        Block(
                            Vector(
                                StarExp(
                                    IdnExp(IdnUse("n")),
                                    AppExp(
                                        IdnExp(IdnUse("fact")),
                                        Vector(
                                            MinusExp(
                                                IdnExp(IdnUse("n")),
                                                IntExp(1))))))))))),
        PrintExp(AppExp(IdnExp(IdnUse("fact")), Vector(IntExp(5))))))

Environment out:
----------------

scope
    "fact" -> Function(fn(int)->int,Block(Vector(IfExp(EqualExp(IdnExp(IdnUse(n)),IntExp(0)),Block(Vector(IntExp(1))),Block(Vector(StarExp(IdnExp(IdnUse(n)),AppExp(IdnExp(IdnUse(fact)),Vector(MinusExp(IdnExp(IdnUse(n)),IntExp(1)))))))))))

Target tree:
------------

List(
    IClosure(
        Some("fact"),
        List("n"),
        List(
            IVar("n"),
            IInt(0),
            IEqual(),
            IBranch(
                List(IInt(1)),
                List(
                    IVar("n"),
                    IVar("n"),
                    IInt(1),
                    ISub(),
                    IVar("fact"),
                    ICall(),
                    IMul())))),
    IClosure(
        None,
        List("fact"),
        List(IInt(5), IVar("fact"), ICall(), IPrint())),
    ICall())

Output:
-------

120
